### PR TITLE
JSON Rest APIs

### PR DESCRIPTION
**Pull request by katonaaron-msg**
- Add product service, mapper, controller and DTOs.

### CODE REVIEW

File: src/main/java/ro/msg/learning/shop/mapper/ProductMapper.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+import java.util.Collection;
+import java.util.List;
+import java.util.UUID;
+
+@Mapper(componentModel = "spring")
+public abstract class ProductMapper {
+
+    @Autowired
+    protected ProductCategoryRepository categoryRepository;
```
Comment (cristinaAnca) [inline src/main/java/ro/msg/learning/shop/mapper/ProductMapper.java]: Do not use repositories in a mapper class. A Mapper should just map data from an entity to a Dto and from Dto to entity.  All the other logic which implies repositories should live outside the mapper, in the service class eventually.

**Review by cristinaAnca — CHANGES_REQUESTED** [review]

File: src/main/java/ro/msg/learning/shop/service/OrderService.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+@Service
+@RequiredArgsConstructor
+public class OrderService {
+
+    private final OrderRepository orderRepository;
+    private final LocationRepository locationRepository;
+    private final AddressMapper addressMapper;
+
+    @Transactional
+    public Order createOrder(CreateOrderDTO createOrderDTO) {
```
Comment (cristinaAnca) [inline src/main/java/ro/msg/learning/shop/service/OrderService.java]: Here you either use Order as a parameter, or you return an OrderDto. But it is not ok for the method to receive a Dto and return an entity (in this case Order). Usually, services work with entities and the controller is the one who does the mapping to Dto. Or there is another option in which the service receives Dto and returns again a Dto and the mapping to the domain is done in the methods inside the service. But there is never a mix between the two options. The preferred option is option 1, where the services works with entities and the controller performs the mapping between entity and dto.

File: src/main/java/ro/msg/learning/shop/service/OrderService.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        final var productIdToQuantity = createOrderDTO.orderDetails().stream()
+                .collect(Collectors.toMap(CreateOrderDetailDTO::productId, CreateOrderDetailDTO::quantity));
+        final var location = locationRepository.findAll().stream()
```
Comment (cristinaAnca) [inline src/main/java/ro/msg/learning/shop/service/OrderService.java]: Can you write a query for this complex operation? It is very expensive to retrieve all the locations first and then filter the wrong ones out. If you have hundreds of locations and only one of these locations is the correct one, you will load all the locations in memory which is an expensive operation.

File: src/main/java/ro/msg/learning/shop/service/OrderService.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                        dto.quantity()
+                ))
+                .collect(Collectors.toSet());
+
+        final var order = new Order();
+        order.setCreatedAt(createOrderDTO.timestamp());
+        order.setOrderDetails(orderDetails);
+        order.setShippedFrom(location);
+        order.setDeliveryAddress(addressMapper.toEntity(createOrderDTO.deliveryAddress()));
+        order.setCustomer(null); //TODO
```
Comment (cristinaAnca) [inline src/main/java/ro/msg/learning/shop/service/OrderService.java]: You need to solve the TODO comment, customer cannot be null

**Review by cristinaAnca — CHANGES_REQUESTED** [review]

File: src/main/java/ro/msg/learning/shop/ShopApplication.java [hunk 5]
```diff
@@ -2,12 +2,35 @@
...
 	}
 
+	/**
+	 * Returns a customer from the database, or a dummy customer if there is none
+	 *
+	 * @param customerRepository
+	 * @return
+	 */
+	@Bean
+	Customer customer(CustomerRepository customerRepository) {
```
Comment (cristinaAnca) [inline src/main/java/ro/msg/learning/shop/ShopApplication.java]: I understand what you want to do with this bean, and it's a good idea. But let's move it elsewhere, in a configuration file, just to be separated from the Application. You have more details here: https://docs.spring.io/spring-javaconfig/docs/1.0.0.m3/reference/html/creating-bean-definitions.html.

**Review by cristinaAnca — CHANGES_REQUESTED** [review]

File: src/main/java/ro/msg/learning/shop/ShopApplication.java (hunk 5 above)
Comment (katonaaron-msg) [reply src/main/java/ro/msg/learning/shop/ShopApplication.java]: I moved it to a separate configuration file :)

**Review by cristinaAnca — APPROVED** [review]

**Pull request merged**